Add more goverance wording to the contribution.md

AloisReitbauer opened the pull request:
Please review, I took standard open source goverance rules and adde them to our contributing md. @plehegar please provide input from a W3C perspective.

AloisReitbauer committed:
Add more goverance wording to the contribution.md

Please review, I took standard open source goverance rules and adde them to our contributing md. @plehegar please provide input from a W3C perspective.

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+
+This document describes the contribution and governance rules for the [W3C Distributed Tracing Working Group](https://www.w3.org/2018/distributed-tracing/). This includes the following specifications
+
+- [Trace Context](https://github.com/w3c/trace-context)
```
why not correlation context?

AloisReitbauer replied:
I can add it. Just did not as we are not technically working on it. I just added it to the spec

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+## Meetings
+In order to provide equitable rights to all members, the following process will be followed:
+
+- Official conference calls are held on a bi-weekly basis and are announce on a [shared calendar](https://calendar.google.com/calendar/embed?src=dynatrace.com_5a09qhua6fh7jb23h7vdjg6veg%40group.calendar.google.com)
```
announce => announced

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+
+- Official conference calls are held on a bi-weekly basis and are announce on a [shared calendar](https://calendar.google.com/calendar/embed?src=dynatrace.com_5a09qhua6fh7jb23h7vdjg6veg%40group.calendar.google.com)
+- Official face-to-face meetings will be announced at least 4 weeks in advance.
+- Proposed changes to any document will be done via a Pull Request (PR).
```
*Suggestion*: formulate it in present time and make accent that this is the only way to make a change. Like "Changes to the document MUST be submitted as Pull Requests after initial discussion in form of GitHub issue or during the meeting."

AloisReitbauer replied:
done

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+- Official conference calls are held on a bi-weekly basis and are announce on a [shared calendar](https://calendar.google.com/calendar/embed?src=dynatrace.com_5a09qhua6fh7jb23h7vdjg6veg%40group.calendar.google.com)
+- Official face-to-face meetings will be announced at least 4 weeks in advance.
+- Proposed changes to any document will be done via a Pull Request (PR).
+- PRs will be reviewed during official meetings, but off-line reviews (LGTMs, NOT LGTMs) and comments are strongly encouraged to gauge the group's opinion on the proposed change prior to the meeting.
```
mention issues for initial discussion as well.

SergeyKanzhelev replied:
There should be process for PRs to be merged without the meeting. Especially for smaller ones

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+- PRs will be reviewed during official meetings, but off-line reviews (LGTMs, NOT LGTMs) and comments are strongly encouraged to gauge the group's opinion on the proposed change prior to the meeting.
+- During meetings, priority will be given to PRs that appear to be ready for a vote over those that appear to require discussions.
+- PRs should not be merged if they have had substantial changes made within two days of the meeting. Rebases, typo fixes, etc. do not count as substantial. Note, administrivia PRs that do not materially modify output documents may be processed by admins as needed.
+- Resolving PRs ("merging" or "closing with no action") will be done as a result of a motion made during a meeting, and approved.
```
Should there be definition of quorum?

mtwo replied:
I think that quorum is defined further down

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+- PRs should not be merged if they have had substantial changes made within two days of the meeting. Rebases, typo fixes, etc. do not count as substantial. Note, administrivia PRs that do not materially modify output documents may be processed by admins as needed.
+- Resolving PRs ("merging" or "closing with no action") will be done as a result of a motion made during a meeting, and approved.
+- Reopening a PR can be done if new information is made available, and a motion to do so is approved.
+- Any motion that does not have "unanimous consent" will result in a formal vote. See Voting.
```
again, there should be a definition of quorum

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+## Participation
+There are three categories of project participation:
+
+1 - Working Group Member. This is anyone who officially participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). Working group members can file PRs against documents and can participate in the PR voting procedure. Working Group members need to be employed by a W3C member organisation or participate as invited experts. 
```
what indices signifies? Is it's a list - should be a dot after index.

SergeyKanzhelev replied:
organisation  -> organization

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+## Participation
+There are three categories of project participation:
+
+1 - Working Group Member. This is anyone who officially participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). Working group members can file PRs against documents and can participate in the PR voting procedure. Working Group members need to be employed by a W3C member organisation or participate as invited experts. 
```
should this "W3C member organisation" be "W3C member organization that officially joined the working group"

SergeyKanzhelev replied:
Also, suggestion I'd lead with this as the beginning of this item suggest general participation when the idea of the item is that you have to join officially

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+There are three categories of project participation:
+
+1 - Working Group Member. This is anyone who officially participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). Working group members can file PRs against documents and can participate in the PR voting procedure. Working Group members need to be employed by a W3C member organisation or participate as invited experts. 
+2 - Non Working Group Member. This is anyone who  participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). In contrast to Working Group members they cannot file a PR or vote on a PR. They can, however, create issues for specifcations. 
```
> cannot file a PR

How can we enforce it and should we?

SergeyKanzhelev replied:
Please install spell checker. "specifcations" -> "specifications"

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+
+1 - Working Group Member. This is anyone who officially participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). Working group members can file PRs against documents and can participate in the PR voting procedure. Working Group members need to be employed by a W3C member organisation or participate as invited experts. 
+2 - Non Working Group Member. This is anyone who  participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). In contrast to Working Group members they cannot file a PR or vote on a PR. They can, however, create issues for specifcations. 
+3 - Chairs and Co-Chairs are members of the group but have the ability to perform administrative actions on behalf of the group. For example, manage the website, github repos and moderate the meetings. Their actions should be done with the knowledge and consent of the group. They also have the ability to merge/close PRs, but only per the group's approval. 
```
I'd suggest to expand on "consent" part of this

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+## PRs
+Typically, PRs are expected to meet the following criteria prior to being merged:
+
+- The author of the PR indicates that it is ready for review by asking for it to be discussed in an upcoming meeting - eg. by adding it to the agenda document.
```
where is agenda document?

SergeyKanzhelev replied:
Can there be some other gesture on PR to suggest readiness. It's unrealistic to expect people to keep track of agenda document before reviewing anything

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+
+- The author of the PR indicates that it is ready for review by asking for it to be discussed in an upcoming meeting - eg. by adding it to the agenda document.
+- All comments have been addressed.
+- PRs that have objections/concerns will be discussed off-line by interested parties. A resolution, updated PR, will be expected from those talks.
```
> off-line

please expand on the definition of off-line here

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+## Voting
+If a vote is taken during a meeting, the follow rules will be followed:
+
+- There is only 1 vote per participating company, or nonaffiliated individual.
```
is it a typical practice? Should the vote of implementer be bigger than party that hasn't implemented? Are only invited nonaffiliated individuals participate? Right now it's only Yuri

mtwo replied:
Adding on to this - does the W3C typically give each committee member a vote or each sponsoring company? We should follow whatever the rest of the W3C does

plehegar replied:
Voting requirements are listed in https://www.w3.org/2018/Process-20180201/#Votes .

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+- Each participating company can assign a primary and secondary representative.
+A participating company attains voting rights by having any of the entity's assigned representative(s), in aggregate, attend 3 out of the last 4 meetings. The people listed as "primary" or "alternate" for an entity can be changed no more than once a month and can be changed by notifying one of the admins. The entity obtains voting rights after the 3rd meeting, not during.
+- Only members with voting rights will be allowed to vote.
+- A vote passes if more than 50% of the votes cast approve the motion.
```
is 50% a typical practice? 70% sounds better for me in terms of guaranteeing better adoption and faireness

mtwo replied:
We should follow whatever the rest of the W3C does

plehegar replied:
W3C Groups rarely conducts formal votes nowadays, so there is no typical number. At the minimum, I suggest adding a quorum requirement before conducting a vote.

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
+
+- There is only 1 vote per participating company, or nonaffiliated individual.
+- Each participating company can assign a primary and secondary representative.
+A participating company attains voting rights by having any of the entity's assigned representative(s), in aggregate, attend 3 out of the last 4 meetings. The people listed as "primary" or "alternate" for an entity can be changed no more than once a month and can be changed by notifying one of the admins. The entity obtains voting rights after the 3rd meeting, not during.
```
I'd suggest to keep a member if participation is online-only. I.e. if company rep actively commenting on PRs and issues and submitting PRs she/he should not be forced to dial in or even fly to the meeting.

yurishkuro replied:
+1 to what @SergeyKanzhelev said, given that it's impossible to find a time for a meeting convenient for every possible member across the globe.

mtwo replied:
+1

plehegar replied:
Our charter doesn't restrict the number of individuals from a participating Member so I can't think we can put a limit outside the charter, per "A group charter may limit the number of individuals representing a W3C Member (or group of related Members)." https://www.w3.org/2018/Process-20180201/#member-rep

SergeyKanzhelev reviewed (changes requested):
I have added a few comments. Process in general feels a little bit heavy. Maybe reasonably heavy, but I'd suggest there should be simplified version of it with simple flow: Issue -> PR -> X approvals with listed name/company affiliation -> Merge. And all the heavy process with casting vote and discussion during the meeting is for the PRs that cannot collect enough approvals. This way we open door for minor contributions and close faster on obvious issues. And allow more time for difficult decision discussion during the meeting.

yurishkuro commented:
> Issue -> PR -> X approvals with listed name/company affiliation -> Merge. And all the heavy process with casting vote and discussion during the meeting is for the PRs that cannot collect enough approvals.

I agree with that. Currently the problem is that so far some PRs have been merged without getting hardly any approvals (i.e. 1). A simple majority vote "+1 binding" / "+1 non-binding" would address my concerns in https://github.com/w3c/trace-context/issues/165 (btw, please move that issue here). If there is a situation when people fail to vote and even simple PRs are getting stuck, then a time frame can be established for "speak now or hold your peace".

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,55 @@
 
-This repository is being used for work in the [W3C Distributed Trace Context Working Group](https://www.w3.org/2018/distributed-tracing/).
 
 ## Editing report
```
btw, this section can be replaced by "see contributing guide of specification repository" as there is no reports in this repository.

AloisReitbauer committed:
updates based on sergeys comments.

yurishkuro commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,56 @@
+- PRs that have objections/concerns will be discussed off-line by interested parties. A resolution, updated PR, will be expected from those talks.
+
+## Voting
+If a vote is taken during a meeting, the follow rules will be followed:
```
-1 to any votes during meeting, instead of explicit votes on PRs or mailing list.

plehegar replied:
-1 as well: our charter says: "To afford asynchronous decisions and organizational deliberation, any resolution (including publication decisions) taken in a face-to-face meeting or teleconference will be considered provisional. A call for consensus (CfC) will be issued for all resolutions (for example, via email and/or web-based survey), with a response period from one week to 10 working days, depending on the chair's evaluation of the group consensus on the issue. If no objections are raised on the mailing list by the end of the response period, the resolution will be considered to have consensus as a resolution of the Working Group. ".

plehegar commented on CONTRIBUTING.md:
```diff
@@ -1,6 +1,56 @@
+## Participation
+There are three categories of project participation:
+
+1. Non Working Group Member. This is anyone who  participates in the group's activities in any of our communication channels (email, github issues/PRs, meetings, etc.). In contrast to Working Group members they can file a PR but not vote on a PR. They can, however, create issues for specifcations. 
```
participation is subject to https://www.w3.org/2018/Process-20180201/#Policies . As such, while we accept and encourage comments and contributions in mailing lists and GH, meeting participants are subject to https://www.w3.org/2018/Process-20180201/#GeneralMeetings .

plehegar commented:
Modulo making sure we don't conflict with the W3C Process and the Working Group charter (see comments), this seems fine to me. I'm worried about the need to do votes however, which is unusual nowadays for a W3C Group. so I'm wondering about the motivation to expand on those here. Otherwise, keep in mind that the W3C Process always prevail, then the charter, then this CONTRIBUTING document. We'll also need to add a link back here from all of the CONTRIBUTING files we have in the spec repos.

yurishkuro commented:
@plehegar on the voting, I see your point, it's probably too heavy process. Instead, there can be a mandatory waiting period, if no objections raised then no vote is required.

AloisReitbauer committed:
merged PR and Voting.

I tried address all concerns

* removed voting not compliant with W3C
* added link to W3C rules
* added reasonable limit for responding to comments.

AloisReitbauer committed:
removed report editing section

removed the report editing section due to Sergey's proposal to move them to the specs.

The change in participatoion is just a typo fix

AloisReitbauer committed:
Changed meeting language

* add recommendation to file an issue of items are not ready for a PR yet
* remove requirement for a formal vote
* change language that PRs are not merge in a meeting  but a work week after

iredelmeier commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,38 @@
+- PRs will be reviewed during official meetings, but off-line reviews (LGTMs, NOT LGTMs) and comments are strongly encouraged to gauge the group's opinion on the proposed change prior to the meeting.
+- If a topic is not defined specifically enough yet to justify a PR against the specification, contributors are encouraged to file an issue. 
+- During meetings, priority will be given to PRs that appear to be ready for a vote over those that appear to require discussions.
+- PRs should not be merged if they have had substantial changes made within two days of the meeting. Rebases, typo fixes, etc. do not count as substantial. Note, administrivia PRs that do not materially modify output documents may be processed by admins as needed.
```
typo: administrivia -> administrative

AloisReitbauer committed:
Changes discussed during yesterdays meeting

* Section on encouraging contributions
* Define PR handling (open for one week)
* Revert of PRs as last resort

AloisReitbauer committed:
adjusted meeting to fit W3c policy

SergeyKanzhelev commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,49 @@
+
+Specificaiton changes are proposed via PRs which are expected to meet the following criteria prior to being merged:
+
+- The author of the PR indicates that it is ready for review by asking for it to be discussed in an upcoming meeting - eg. by adding it to the agenda document.
```
Perhaps there might be another form of indication like remove `[WIP]` from the title?

SergeyKanzhelev commented:
@AloisReitbauer one minor feedback on how to indicate that PR is ready for review. Perhaps this should be assumed unless there is a `[WIP]` in the title?

SergeyKanzhelev commented:
should this be merged?

AloisReitbauer commented:
Merging as there are no open comments and it has been open for a while.

AloisReitbauer committed:
minor wording changes
